docs/markdown/Builtin-options.md

```diff
@@ -76,6 +76,7 @@ machine](#specifying-options-per-machine) section for details.
 | -------------------------------------- | ------------- | -----------                                                    | -------------- | ----------------- |
 | auto_features {enabled, disabled, auto} | auto       | Override value of all 'auto' features                          | no             | no                |
 | backend {ninja, vs,<br>vs2010, vs2012, vs2013, vs2015, vs2017, vs2019, vs2022, xcode} | ninja | Backend to use        | no             | no                |
+| genvslite {vs2022}                     | vs2022        | Setup multiple buildtype-suffixed ninja-backend build directories (e.g. [builddir]_[debug/release/etc.]) and generate [builddir]_vs containing a Visual Studio solution with multiple configurations that invoke a meson compile of the newly setup build directories, as appropriate for the current configuration (builtype) | no | no |
 | buildtype {plain, debug,<br>debugoptimized, release, minsize, custom} | debug |  Build type to use                    | no             | no                |
```

Comment:
We typically want to avoid adding new options whenever possible. In this case could this instead be implemented by adding a new value for `backend`? Something like `vslite`?

Reply:
@jpakkane I think I may have thought about that as a first attempt but decided against it because I'd argue that this new functionality doesn't behave in the same way as the other 'backend' options;  all other backends create only one build directory that configures the build for a single, specific buildtype.  I'd argue most reasonable users would expect just another 'backend' option would lead most reasonable users to expect the same pattern of setup behaviour and build directory creation... Not that it instead goes and creates a set of (currently) 4 suffixed build directories, for different buildtypes, and one for the generated solution and project files to tie it all together.

Reply:
That may be so, but the backend is currently not an "option". It can not be changed after the build dir is set up. This however is an option so people would expect to be able to change between the two types by changing the option. That is not to be accepted, the "kindness" of a build dir needs to be immutable.

Since we probably need to make some backwards compatibility breaking changes to this before it is complete, maybe the type should be "exp-vslite" instead of "vslite"? (I considered "unstable-vslite", but it's a bit too much to type IMHO.).

Reply:
Sorry but I don't follow most of what you're trying to say.  Can you be more be more fully qualified or throw in examples of what you think would be a problem and what you think a better solution would be?

> backend is currently not an option

Do you mean "the --backend option would currently have no effect in conjunction with the --genvslite option"?  If so, that's correct;  as stated by the usage help/description -

> Setup multiple buildtype-suffixed ninja-backend build directories (e.g. builddir_[debug/release/etc.]) and generate [builddir]_vs containing a Visual Studio solution ...


> people would expect to be able to change between the two types by changing the option.

What two types are you referring to here?

Are you suggesting someone would do `meson setup --genvslite vs2022 somebuilddir somesrcdir` to set up somebuilddir_debug/debugoptimized/release/vs and then want to reconfigure some of the setup settings in one or more of these multiple buildtype-suffixed build dirs?  If so, how is that a problem?  I think it's clear to the users that the --genvslite option acts as a kind of 'setup macro' for generating multiple normal, ninja-backended build directories, so if they then really want to go and manually tweak/reconfigure what has been set up in their 'somebuilddir_release' directory, for example, then that's up to them and they should (and I think can) understand what they'll get.

> we probably need to make some backwards compatibility breaking changes to this before it is complete

Can you explain further, please? In this change's current state, nothing of this new option breaks any of the previous behaviours that I'm aware of.  Are you suggesting making changes that would cause back compat. breaking changes?  If so, can you go into more detail, please?

Reply:
> Are you suggesting someone would do meson setup --genvslite vs2022 somebuilddir somesrcdir and then want to reconfigure some of the setup settings in one or more of these multiple buildtype-suffixed build dirs?

What I'm saying is that if people set up a vs backend with `genvslite` set to `true`, then they expect that if you change that to `false` later and reconfigure, their build dir changes from "lite style" to "classical style" VS project. And that is something we _don't_ want to support because mixing different styles of project styles within a single directory seems extremely fragile.

So what I'm suggesting is that this would be set up with something like `meson setup --exp-vslite somebuilddir somesrcdir`. Or maybe it should be called `exp-vsninja` to make it more clear what it does. But the point is that once the build dir is set up to contain a "lite" vs project, it _can not change_. Because of that the setting can't really be a Meson option.

> Can you explain further, please? In this change's current state, nothing of this new option breaks any of the previous behaviours that I'm aware of

What we have discovered through experience is that whenever new functionality like this is added, some part of it goes wrong. Not because of the commit would be bad, but because in real usage we find some unforeseen problems or quirks that would require breaking changes. As this is a totally new kind of a backend, I suspect we will find run into these. This is why it would be nice to keep this MR as isolated as possible and also mark it explicitly as unstable so we can change it later (and if people complain, then that is their problem). We do the same for new modules, they are typically first marked as `unstable-modname`.

Reply:
> That may be so, but the backend is currently not an "option". It can not be changed after the build dir is set up. This however is an option so people would expect to be able to change between the two types by changing the option. That is not to be accepted, the "kindness" of a build dir needs to be immutable.

I regret to inform you both that it's currently an option, and that you can currently change it during --reconfigure.

Now granted, if you do it will fully configure the entire thing, then traceback at the last second, see #10014 -- but that's only because of backend-specific options, you cannot move between vs and ninja, and I think you *can* move from ninja *to* xcode.

Reply:
> What I'm saying is that if people set up a vs backend with genvslite ...

This is still a little confusing and I'm wondering if there's a fundamental misunderstanding or assumption or terminiology that means different things to us.  Perhaps others could shed some light if you can see the possible source of miscommunication.  The above quoted sentence does not make any sense to me;  use of genvslite cannot set up a vs backend;  it effectively forces the --backend ninja option for multiple different (suffixed) build directories.  What will happen, if someone does `meson setup --genvslite vs2022 somebuilddir somesrcdir`, is it will (internally) go off and run the following setup steps -
- `meson setup --backend ninja --buildtype debug somebuilddir_debug somesrcdir`
- `meson setup --backend ninja --buildtype debugoptimized somebuilddir_debugoptimized somesrcdir`
- `meson setup --backend ninja --buildtype release somebuilddir_release somesrcdir`
- And also generate a directory called 'somebuilddir_vs' containing the .sln and .vcxproj files that glues these all together in a multi-configuration (a.k.a 'buildtype') solution.

It is very much just like a meson setup macro and is pretty unambiguous about what it's doing, what build directories you'll end up with, and how they're configured.  Is the fact that you end up with multiple buldtype-suffixed build directories (which is unique to this setup option) really at the heart of your concerns here?

> ... if you change that to false later and reconfigure, their build dir changes from "lite style" to "classical style" VS project ... that is something we don't want to support because mixing different styles of project styles within a single directory seems extremely fragile.

To me, this seems like quite a strange example that looks like a misunderstanding of what the --genvslite setup option will do.  If, after someone does `meson setup --genvslite vs2022 somebuilddir somesrcdir`, then they want to change back to a 'classical style VS project', then that means they want to go back to a single build directory, containing a single buildtype-configured .sln and .vcxprojs (that will build with the msbuild machinery).  To do so would mean they almost certainly want to set that up in a brand new 'somebuilddir', not the recently generated 'somebuilddir_[debug/debugoptimized/release/vs]'.  However, if they really wanted (and though I can't think of a realistic scenario in which someone would want to do this), there's nothing stopping them from piggybacking and reconfiguring one of the previously setup buildtype-suffixed build directories.  E.g. `meson setup --reconfigure -Dsomeoption=somenewval --backend vs --buildtype debug somebuilddir_debug somesrcdir` so that 'somebuilddir_debug' is then reconfigured from its previous ninja-backend to the msbuild stuff.
If they were to try `meson setup --reconfigure -Dsomeoption=somenewval --backend vs --buildtype debug somebuilddir somesrcdir` (note the build directory name difference), then that wouldn't work or (in my opinion) be expected to work because 'somebuilddir' doesn't exist.

> So what I'm suggesting is that this would be set up with something like `meson setup --exp-vslite somebuilddir somesrcdir`

I think this just confirms to me that either I or you (or both) misunderstand something, since your suggestion appears to be essentially the same, only changing the name of the setup option from 'genvslite' to 'exp-vslite', as well as getting rid of the set of supported visual studio versions for which the user wants to generate the .sln/.vcxproj files.  Regarding the enumeration of 'vsxxxx' versions:  Sure, currently, I only had the option to generate for vs2022 but there's nothing stopping the addition of supporting generation of older or newer versions, which will need some way of specifiying the generate .sln/.vcxproj versions.  What am I missing here?

While writing this, I've been wondering if perhaps you're expectation is that `meson setup --genvslite vs2022 somebuilddir somesrcdir` will (or should) still build a SINGLE build directory called 'somebuilddir', and then within that single 'somebuilddir' will be further hidden, internal build directories (e.g. 'somebuilddir/debug', 'somebuilddir/debugoptimized', 'somebuilddir/release') that are the conventional meson ninja build dirs.  I think doing something like that would make things a bit more complicated and perhaps less intuitive.  The previous "multi-config-suffixed build dirs" behaviour (i.e. the current genvslite behaviour in this PR) makes it about as clear and obvious as I think is possible that you're getting multiple buildtype configured ninja build dirs.  Any attempt to shoehorn everything in to a single build dir now introduces a bit of ambiguity as to what kind of buildtype (debug/opt/release/etc) the one builddir is configured as well as tries to bend over backwards to make everything continue to look like the conventional single builddir, single configuration approach when that's no longer the case..  And since the whole point of the option is to support a multi-config/buildtype VS solution and project files that still build through meson's ninja back-end, what does it mean for someone to try to setup or '--reconfigure' their one builddir for a different buildtype?  With this PR's proposed approach, I'd argue it's just simple in that it ignores any '--builtype X' in the setup command line and just plops out '..._debug', '..._debugoptimized', and '..._release' configured ninja-backended build directories, and that an approach to somehow squeeze this multi-buildtype setup functionality into a single build directory feels like too much of a contortion and raises more questions and increases complexity over this current 'multiple suffixed build dir setup macro'-like functionality.

Reply:
> I regret to inform you both that it's currently an option, and that you can currently change it during --reconfigure.

[That is a bug that will be fixed](https://github.com/mesonbuild/meson/pull/11571).

Reply:
> argue it's just simple in that it ignores any '--builtype X' in the setup command line and just plops out '..._debug', '..._debugoptimized', and '..._release' configured ninja-backended build directories

What you could do is something like:

````
build
build\top_level_vs_solution.sln
build\debug
build\debugoptimized
...
````

In this way everything is still contained in a single build dir for easy deletion but the separate build types have their own dirs.

Reply:
> What you could do is something like ... In this way everything is still contained in a single build dir for easy deletion...

This is just what I said a couple of posts above :confused: -

> I've been wondering if perhaps you're expectation is that meson setup --genvslite vs2022 somebuilddir somesrcdir will (or should) still build a SINGLE build directory ... and then within that single 'somebuilddir' will be further hidden, internal build directories ...

And I go on to explain why I think doing that is a bit more complicated and problematic than this solution.

I think having multiple build dirs concurrently is fine and identifying or deleting them, just as for deleting any single build dir, is something I can't imagine is troublesome for anyone.